jax/numpy/lax_numpy.py

```diff
@@ -1300,6 +1300,32 @@ def select(condlist, choicelist, default=0):
   return output
 
 
+@_wraps(onp.bincount, lax_description="""\
+Jax adds the optional `length` parameter which specifies the output length, and
+defaults to ``x.max() + 1``. It must be specified for bincount to be compilable.
+""")
+def bincount(x, weights=None, minlength=0, *, length=None):
+  if length is None:
+    length = x.max() + 1
+  length = _max(length, minlength)
+  x = array(x)
+  if ndim(x) != 1:
+    raise ValueError("only 1-dimensional input supported.")
+  if weights is None:
+    dtype='int32'
+    def update(i, result):
+      return ops.index_add(result, x[i], 1)
+  else:
+    if shape(x) != shape(weights):
+      raise ValueError("shape of weights must match shape of x.")
+    weights = array(weights)
+    dtype=weights.dtype
+    def update(i, result):
+      return ops.index_add(result, x[i], weights[i])
+  result = zeros(length, dtype=dtype)
+  return lax.fori_loop(0, x.shape[0], update, result)
```

Comment:
Can't we avoid the loop by using advanced indexing?

e.g., consider the example from the numpy docs:
```
>>> w = np.array([0.3, 0.5, 0.2, 0.7, 1., -0.6]) # weights
>>> x = np.array([0, 1, 1, 2, 2, 2])
>>> np.bincount(x,  weights=w)
array([0.3, 0.7, 1.1])
```

```
>>> jax.ops.index_add(jnp.zeros((3,), jnp.float32), jax.ops.index[x], w)
array([0.3, 0.7, 1.1])
```

In general you'll do much better by emitting a single `scatter` operation rather than a loop of individual updates, since it allows the backend freedom to choose a better implementation strategy. e.g., on GPU the scatter-add will be parallelized.

Reply:
That's a great idea – I didn't know those operations were available... this looks like it's essentially the equivalent of ``np.ufunc.at``.